### DISCUSSION TITLE
Rust Implementation

### DISCUSSION THREAD

**Discussion by aspizu** [apple/pkl#49] [General]
Hey, I want to try writing a rust implementation. Could you explain the parts of the core?

**Comment by holzensp** [comment]
Are you talking about language bindings for Rust, or a re-implementation of the language in Rust? If the former; take a look at `pkl-swift` or `pkl-go`. If the latter; please ask a few more directed questions.

**Reply by aspizu** [reply]
The latter please, could you explain the steps required to run a pkl program? like parsing and type-checking.

**Reply by holzensp** [reply]
I feel you maybe underestimating what building a language interpreter involves. A deep dive on how Pkl does all of its evaluation is sufficient for a significant book. If you have questions about specific parts of the code, I'm happy to help, but I can't give you ten bullet points to send you on your way for this.

I can tell you we parse using [ANTLR](https://www.antlr.org) (see the [lexer](https://github.com/apple/pkl/blob/main/pkl-core/src/main/antlr/PklLexer.g4) and [parser](https://github.com/apple/pkl/blob/main/pkl-core/src/main/antlr/PklParser.g4) definitions) and we use [Truffle](https://www.graalvm.org/latest/graalvm-as-a-platform/language-implementation-framework/) to define our evaluation.

**Reply by bioball** [reply]
I think it makes sense to provide full specification once we have a 1.0 release out. Until the, parts of the language and syntax are likely to change. I'd maybe hold off on re-implementing Pkl until then.

**Reply by adaflexer** [reply]
Using the lexer and parser, we could parse pkl files via Python?

**Reply by holzensp** [reply]
To what end?

**Reply by aspizu** [reply]
I use LALRPOP and Logos to lex and parse code. Both being Rust libraries, not sure what you mean by 'via Python'?

**Reply by adaflexer** [reply]
We want to parse the PKL files and transform them to another special format (XML like). But we have python developers only. For parsing the created PKL files, and transforming we would like to use python.

**Reply by aspizu** [reply]
what you want is to generate XML by running pkl code, not by parsing it.
I don't know if they have a XML generator yet, but if they don't if you can figure out how to convert JSON to the format of your liking, then you could first run pkl and generate JSON then convert it.

**Reply by adaflexer** [reply]
No, not by running (wording running is ok too btw), we have a special XML format which is not supported by PKL. We want to transform PKL files to this format. To do this we want to be able to parse the PKL files. 
Example PKL.org
name = "Swallow"

job {
  title = "Sr. Nest Maker"
  company = "Nests R Us"
  yearsOfExperience = 2
}

I want to acces to this content and parse the informations out of it and output another file, which is some special XML file. Means I need to be able to add some own logic in to generator / output file.

**Reply by aspizu** [reply]
pkl is a programming language, you need to run it.

**Reply by adaflexer** [reply]
Running / Compiling is ok. We want to just have different output, using python. The question is can we use the g4 files in combination with this https://pypi.org/project/antlr4-python3-runtime/ or I m missing something?

**Reply by aspizu** [reply]
there is no reason to parse pkl at all, do what i said, first generate JSON, then convert JSON to whatever format you like.

**Reply by adaflexer** [reply]
https://github.com/apple/pkl/tree/main/pkl-core/src/main/java/org/pkl/core/parser  
I need this files in python. This is all I m asking, we tryed to generate python parser with given g4 files but seen some errors. Maybe I m missing something regarding antlr.

**Reply by aspizu** [reply]
you won't be able to generate python parsers using the antlr grammer, there is Java code embedded in it.

**Reply by holzensp** [reply]
@aspizu is saying sensible things. There is _a lot more_ going on than just _parsing_ when you want the data out of a Pkl file. But also, rewriting data from one format to another is exactly Pkl's core strength. I think that if Pkl can't produce your XML format, your format isn't actual XML (but I may be wrong). If it's normal XML, but following a specific schema, I'd also do the rewrite in Pkl. Worst-case, I'd let Pkl generate its standard XML output and XSLT to transform it to the target I want.

**Reply by adaflexer** [reply]
Thank will start to check everything and consider transforming to YAML /JSON. We have already the generators written in python. It takes input from textx based DSL Files and transforms to Schema based XML File. We want to try and see if we can use PKL and eliminate textx based language.

**Reply by aspizu** [reply]
Does pkl perform any kind of JIT or compilation?

**Reply by holzensp** [reply]
It piggy-backs on the JIT facilities of the JVM when run on the JVM. For the binary, it uses GraalVM's AOT compiler to sort out optimisations.

**Reply by bioball** [reply]
Pkl code itself is also JIT-compiled, thanks to being implemented in Truffle. Truffle (and GraalVM) is some pretty neat technology. Here's a video about Truffle in case you wanted to learn more:

https://www.youtube.com/watch?v=SYg4KO8Z2Ts

**Comment by zaxebo1** [comment]
@aspizu 
pkl is a great configuration language.
 And your idea of reimplementing it fully in rust is even-greater. 
thousand Github stars to your future repo in advance from my side.

**Comment by akdor1154** [comment]
I did go looking for a rust implementation, not for any special love for rust, but because what I really want is a wasm implementation I could run in-browser. If the native builds could output a wasm target I'd be over the moon, but it sounds like this could be fiddlier than it sounds?

**Reply by aspizu** [reply]
A rust implementation would be wonderful as its very easy to compile to WASM, I'm doing it for my other compiler project and its working great.

**Reply by odenix** [reply]
If you’d like to see a Pkl WASM implementation, make sure to upvote this issue: https://github.com/oracle/graal/issues/3391

**Comment by sandstrom** [comment]
@holzensp I know it's a large undertaking (but easier today with LLMs), but are there any plans to re-implement pkl in Rust?

When it comes to language bindings, most other programming languages usually have good facilities for writing bindings to either C/C++ or Rust code.

For example, I looked a bit at bindings for Ruby, and it's probably doable, but would have been easier if pkl was written in C/C++ or Rust.

**Comment by donbeave** [comment]
I have not seen this repository mentioned in the thread, so I wanted to add it here:

- Repository: [`jdx/pklr`](https://github.com/jdx/pklr)
- Release reviewed: [`v1.3.0`](https://github.com/jdx/pklr/tree/v1.3.0)
- Description: a pure-Rust Pkl parser and evaluator

This appears to be a **real independent implementation**, rather than Rust bindings around the official evaluator or a wrapper that shells out to the `pkl` executable.

The crate contains its own lexer, parser, evaluator, runtime value representation, import resolution, resource access abstraction, and diagnostics. Its primary API evaluates a `.pkl` module directly into a `serde_json::Value`, without requiring an external Pkl installation.

### What is already implemented

Based on the repository, tests, and changelog, `pklr` already covers a substantial amount of the language and evaluator:

- lexer, parser, and evaluator written in Rust;
- local, HTTP, and selected `package://` imports;
- `amends` and `extends`;
- classes, inheritance, `super`, `this`, and `outer`;
- open classes and property modifiers such as `hidden`, `const`, and `abstract`;
- object amendment and late binding;
- type aliases, constrained types, and `is`/`as`;
- annotations;
- string interpolation and raw strings;
- lambdas and higher-order collection methods;
- glob imports;
- `read()` and `read?()`;
- durations, data sizes, `NaN`, and `Infinity`;
- output renderer converters;
- parser-only and capability-backed builds;
- rich error diagnostics.

The changelog shows that it is actively being developed and is already used as the default embedded evaluator by [`hk`](https://hk.jdx.dev/pkl_introduction.html#dependencies).

Therefore, calling it only a parser, an early experiment, or a thin partial implementation would be unfair.

### Is it a complete Pkl implementation?

My conclusion is:

> **`pklr` is a substantial native Rust reimplementation of Pkl's parser and a broad evaluator, but it is not yet a fully Pkl-compatible runtime or a drop-in replacement for the complete official Pkl implementation and toolchain.**

There are several reasons I would avoid describing it as “complete.”

#### Its own ecosystem documentation says compatibility is incomplete

The [`hk` documentation](https://hk.jdx.dev/pkl_introduction.html#dependencies) says that `pklr` may not yet support every Pkl feature and recommends falling back to the official CLI.

**Reply by aspizu** [reply]
shutup
